src/main/java/com/example/MyBatisDemo/controller/StudentController.java

```diff
@@ -15,33 +23,30 @@
 public class StudentController {
 
     @Autowired
-    StudentRepository studentRepository;
-
+    StudentService studentService;
 
     @GetMapping
-    public List<StudentEntity> getAllStudents(){
-        return studentRepository.findAllStudents();
+    public List<StudentDisplayDto> getAllStudents(){
+        return studentService.getAllStudents();
     }
 
     @GetMapping("/{id}")
-    public StudentEntity getStudentById(@PathVariable(value = "id")Long id){
-        return studentRepository.findStudentById(id);
+    public StudentDisplayDto getStudentById(@PathVariable(value = "id")Long id) {
+        return studentService.getStudentById(id);
     }
 
     @GetMapping("/{id}/subjects")
-    public StudentEntity findSubjectsByStudentId(@PathVariable(value = "id")Long id){
-        return studentRepository.findSubjectsByStudentId(id);
+    public StudentSubjectsDisplayDto findSubjectsByStudentId(@PathVariable(value = "id")Long id) {
+        return studentService.findSubjectsByStudentId(id);
     }
 
     @PostMapping
-    public StudentEntity addStudent(@RequestBody StudentEntity student){
-        Long id = studentRepository.addStudent(student);
-        return studentRepository.findStudentById(id);
+    public StudentDisplayDto addStudent(@Valid @RequestBody StudentAdditionDto studentAdditionDto) {
+        return studentService.addStudent(studentAdditionDto);
     }
 
     @PostMapping("/{id}/subjects")
-    public StudentEntity assignSubjectsToStudent(@PathVariable(value = "id")Long id, @RequestBody List<Long> subjectIds){
-        studentRepository.assignSubjectsToStudent(id,subjectIds);
-        return studentRepository.findSubjectsByStudentId(id);
+    public StudentSubjectsDisplayDto assignSubjectsToStudent(@Valid@PathVariable(value = "id")Long id, @RequestBody StudentSubjectsAdditionDto subjectsAdditionDto) {
+        return studentService.assignSubjectsToStudent(id, subjectsAdditionDto);
```

Comment:
We will also have to make sure that the SubjectAdditionDto does not have duplicate Subject Ids.

Reply:
it is implemented in services.